### DISCUSSION TITLE
Launch OBS Scripts

### DISCUSSION THREAD

**Discussion by gdhgdhgdh** [cpyarger/obs-midi#112] [Ideas]
Hello!

I was amazed to find this project and hoped that I'd be able to launch Lua scripts by sending a MIDI message.

I use multiple copies of the [transformer](https://github.com/bromagosa/obs-transformer/blob/master/transformer.lua) script to trigger on-screen animations and I'd love to be able to assign some of these to my Novation Launchpad Mini.

I looked through the actions but couldn't see anything relating to triggering OBS Scripts - is this something you'd consider adding to `obs-midi` ?

Cheers,
Gavin.

**Comment by lorado** [comment]
I also just found this library.
I also would like to have the ability to run OBS Scripts via MIDI Controller (currently using Stream Deck MIDI Button plugin, which controls virtual MIDI device).
I am using modified instant replay script, just for info =)

**Comment by lorado** [comment]
Hi @cpyarger I was wondered, you closed the issue, but became happy again to see, that issue was converted to discussion =)

A side node - I dig a little bit in the code yesterday. If I got it right, your code is inspired by obs-websockets plugin (or is probably even a fork). I am trying to add the functionality by myself. I wrote some code already (still some bugs), but I am struggling with build =\ 

I am on macOs, it breaks with fatal error, that `jack/jack.h` is not found (but I already installed this...) Anyway, gonna try solve this.

As I come from web-development, I am not really confident with C++, so I don't really know any gotchas (like memory leaks or similar), so I am using code you already have, copy-paste and adopt it.

Gonna send a PR, when I get something working

**Reply by cpyarger** [reply]
Alzy and i heavily pulled from obs-websocket at the start.

**Comment by lorado** [comment]
I still can't build on my local machine... I am building it in GitHub Actions for now, it works for me :D 
And I also got the functionality - running hotkey function on MIDI input.

Here is small preview of what I did. There is new action called `Trigger_Hotkey_By_Name`. In the `Hotkey` dropdown all registered hotkeys in OBS are available, which are registered via `obs_hotkey_register_frontend()`.

I am working on fixing bugs right now, and gonna try to optimize code a little bit. And probably will try to display a description names of hotkeys, instead of hotkey names


<img width="1001" alt="Bildschirmfoto 2021-03-13 um 13 02 52" src="https://user-images.githubusercontent.com/4480983/111029340-89b0e180-83fc-11eb-8fcd-e195147f2543.png">
<img width="400" alt="Bildschirmfoto 2021-03-13 um 13 03 00" src="https://user-images.githubusercontent.com/4480983/111029352-97fefd80-83fc-11eb-8a30-2455d1ee93b8.png">

**Comment by lorado** [comment]
I made a PR, it works. #115 
@gdhgdhgdh if you would like to test it out, here are builds for Windows and mac:
https://www.dropbox.com/s/nkq9mhwuv52v2wc/obs-midi-addjusted-by-lorado.zip?dl=0

Linux build didn't run in GitHub Actions, because of this weird error:
```
E: The repository 'https://download.opensuse.org/repositories/devel:/kubic:/libcontainers:/stable/xUbuntu_20.04  Release' no longer has a Release file. 
```

**Comment by gdhgdhgdh** [comment]
Nice nice work :+1:  - I actually got the Python + obs-websocket one working OK, and I use Linux so I'll have to figure out how to build this. I'll try to make some time for that tomorrow. Today is too busy, I'm afraid.

**Reply by lorado** [reply]
Hi again. Christopher made a good improvement to CI, so artifacts are also created on pull requests, so you are able to download prebuilt package here: 
https://github.com/cpyarger/obs-midi/actions/runs/649653308

There is now also linux build, so you don't have to struggle with building

**Comment by cpyarger** [comment]
So far It looks good, I am asking around about getting those hotkeys translated. (: please also add a line in data/locale/en-us.ini  for translating your new action name.

**Comment by cpyarger** [comment]
Fixed the CI, and added the translation string, You will need to delete your config now that the action name has changed.
